### ISSUE TITLE
Constraints stopped working after adding of some dependancies

### ISSUE THREAD

**Issue by demon101** [grails/grails-data-mapping#1145]
I have a really strange bug.

In one domain contraints configuration sometimes not working.
I mean, in constraint defined **field nullable: true**, but during saving I getting message **on field 'field': rejected value [null];**

It happens on some deploys. After redeploying of one war file to tomcat. After 20% of deployments I'm getting the bug.

Bug has been reproduced in code like:

```groovy
@Transactional
def testCreate(){
    def wp = new Domain(date: new Date()).save(failOnError:true, flush:true)
    redirect(action: 'index')
}
```
I'm really discouraged. 
There were no changes in the domain for many days. Other domains in the project not affected. 

Hosted on AWS. tomcat 8.0.50. Can't reproduce locally. We didn't change a configuration of server for months 

Following dependencies has been added during last month:

```
 compile("org.springframework.boot:spring-boot-starter-integration")
 compile("org.springframework.integration:spring-integration-jdbc")
 compile("org.springframework.integration:spring-integration-http")
 compile "org.grails.plugins:spring-security-rest:2.0.0.M2"
```
   
I really need help, for understanding, where to dig

Grails 3.2.11
gorm 6.0.12.RELEASE

**Comment by jeffbrown** [comment]
The error message you showed there doesn't indicate which constraint failed.  Is it `nullable` that is failing?

**Comment by demon101** [comment]
@jeffbrown yes, in my app I'm getting **rejected value [null];** message for all nullable fields of one domain  with empty values. Looks like constraints section ignored
```
static constraints = {
        field1 nullable: true
        field2 nullable: true
```

It works well before, but somehow It was broken.
Also, it not reproducible in tests

**Comment by demon101** [comment]
As I can see, spring-security-rest tugs many dependancies. Can it be a reason?
> +--- org.grails.plugins:spring-security-rest:2.0.0.M2
> |    +--- com.google.guava:guava-io:r03
> |    |    +--- com.google.guava:guava-annotations:r03
> |    |    \--- com.google.guava:guava-base:r03
> |    |         \--- com.google.guava:guava-annotations:r03
> |    +--- org.pac4j:pac4j-core:1.8.3
> |    |    \--- org.slf4j:slf4j-api:1.7.12 -> 1.7.25
> |    +--- org.pac4j:pac4j-oauth:1.8.3
> |    |    +--- org.pac4j:pac4j-core:1.8.3 (*)
> |    |    +--- org.apache.commons:commons-lang3:3.1 -> 3.4
> |    |    +--- commons-codec:commons-codec:1.10
> |    |    +--- org.scribe:scribe:1.3.6
> |    |    \--- com.fasterxml.jackson.core:jackson-databind:2.5.1 -> 2.8.8 (*)
> |    +--- org.pac4j:pac4j-cas:1.8.3
> |    |    +--- org.pac4j:pac4j-core:1.8.3 (*)
> |    |    +--- org.jasig.cas.client:cas-client-core:3.4.1
> |    |    |    \--- org.slf4j:slf4j-api:1.7.1 -> 1.7.25
> |    |    \--- org.jasig.cas.client:cas-client-support-saml:3.4.1
> |    |         +--- org.jasig.cas.client:cas-client-core:3.4.1 (*)
> |    |         +--- joda-time:joda-time:2.7 -> 2.9.9
> |    |         \--- org.slf4j:slf4j-api:1.7.1 -> 1.7.25
> |    +--- com.nimbusds:nimbus-jose-jwt:4.8
> |    |    +--- net.jcip:jcip-annotations:1.0
> |    |    +--- net.minidev:json-smart:1.3.1
> |    |    \--- commons-io:commons-io:2.4
> |    +--- com.lowagie:itext:2.0.8 -> 2.1.7
> |    +--- commons-codec:commons-codec:1.10
> |    \--- org.grails.plugins:spring-security-core:3.0.2 -> 3.1.2 (*)
>

**Comment by demon101** [comment]
I have found interesting bug:
https://stackoverflow.com/questions/43227084/grails-unit-test-null-strangeness

Is it possible to clean constraints of domain by bindData not in test?

**Comment by jeffbrown** [comment]
> Is it possible to clean constraints of domain by bindData not in test?

I don't understand the question.

**Comment by demon101** [comment]
@jeffbrown the bug looks like, all settings from constraints section gone. Can it be erased in run time? 
I see fresh line in the code
` bindData(domain, command)`
Now I'm already building version for test with 
` bindData(domain, command, [exclude: ['constraints', 'hasMany', 'mapping']]) `

**Comment by jeffbrown** [comment]
FYI... You should not have to `exclude: ['constraints', 'hasMany', 'mapping']`.  Static members don't participate in data binding.

**Comment by demon101** [comment]
I need more time for tests.
But by the question, https://stackoverflow.com/questions/43227084/grails-unit-test-null-strangeness looks like it can happen during binding command with constraints to entity. I have the same case.

**Comment by demon101** [comment]
@jeffbrown heh, in one place I have found code
bindData(domain, command, [exclude:['class']])

I don't know why class excluded, and I can't ask the author, he is on vacation now, but looks like bindData sometimes provides surprises =)

**Comment by jeffbrown** [comment]
> But by the question, https://stackoverflow.com/questions/43227084/grails-unit-test-null-strangeness looks like it can happen during binding command with constraints to entity. I have the same case.

The accepted answer there points out that `User.constraints` is `null`, which is misleading because the value of `User.constraints` isn't relevant at that point.  The validator doesn't look at `User.constraints`.  The `constraints` block is evaluated early in the process and the result of that is the validation constraints are configured.

**Comment by demon101** [comment]
@jeffbrown I have reproduced coping of static fields in integration test of our app.
Will provide test app soon

**Comment by demon101** [comment]
@jeffbrown Hi,
I have created a small example app. Sorry for a bad structure of test, I spent hours for cropping code around. You can find code in ApiController

https://github.com/demon101/copy-constraints-bug

./gradlew integrationTest fails with error
 
```
org.grails.datastore.mapping.validation.ValidationErrors: 2 errors
Field error in object 'copy.bug.WorkingPeriod' on field 'class': rejected value [null]; codes [copy.bug.WorkingPeriod.class.typeMismatch.error,copy.bug.WorkingPeriod.class.typeMismatch,workingPeriod.class.typeMismatch.error,workingPeriod.class.typeMismatch,typeMismatch.copy.bug.WorkingPeriod.class,typeMismatch.class,typeMismatch.java.lang.Class,typeMismatch]; arguments [class]; default message [Cannot set readonly property: class for class: copy.bug.WorkingPeriod]
Field error in object 'copy.bug.WorkingPeriod' on field 'note': rejected value [null]; codes [copy.bug.WorkingPeriod.note.nullable.error.copy.bug.WorkingPeriod.note,copy.bug.WorkingPeriod.note.nullable.error.note,copy.bug.WorkingPeriod.note.nullable.error.java.lang.String,copy.bug.WorkingPeriod.note.nullable.error,workingPeriod.note.nullable.error.copy.bug.WorkingPeriod.note,workingPeriod.note.nullable.error.note,workingPeriod.note.nullable.error.java.lang.String,workingPeriod.note.nullable.error,copy.bug.WorkingPeriod.note.nullable.copy.bug.WorkingPeriod.note,copy.bug.WorkingPeriod.note.nullable.note,copy.bug.WorkingPeriod.note.nullable.java.lang.String,copy.bug.WorkingPeriod.note.nullable,workingPeriod.note.nullable.copy.bug.WorkingPeriod.note,workingPeriod.note.nullable.note,workingPeriod.note.nullable.java.lang.String,workingPeriod.note.nullable,nullable.copy.bug.WorkingPeriod.note,nullable.note,nullable.java.lang.String,nullable]; arguments [note,class copy.bug.WorkingPeriod]; default message [Property [{0}] of class [{1}] cannot be null]
	at copy.bug.ApiController.$tt__save(ApiController.groovy:21)
```

on line 
```groovy
//        bindData(period, command, [exclude: ['hours','class', 'constraints']]) //works well
//        bindData(period, command) //works well
        bindData(period, command, [exclude: ['hours']]) // fails during saving

```

class -- native field
note -- has nullable: true

**Comment by jeffbrown** [comment]
I have not verified but I expect the problem here is that `class` would normally be excluded by default.  Since you are providing your own `exclude` list, that takes the place of the the default exclusion list so `class` isn't being ignored.

**Comment by demon101** [comment]
I hope It can be fixed in same way with https://github.com/grails/grails-core/issues/11054
by using LazyMetaPropertyMap instead of LazyBeanMap in DefaultDataBindingSourceCreator

Will test tomorrow, I hope, I will find time for it

**Comment by demon101** [comment]
@jeffbrown I think, fix for grails/grails-core#11054 not resolve the bug. It's another direction, command -> domain. 
But I will try to fix it too, I hope I will find a time this week. I had terrible release yesterday

**Comment by demon101** [comment]
fix merged, I hope will be shipped in 3.3.10
